### PR TITLE
[FLINK-25188][python][build] Support m1 chip.

### PR DESCRIPTION
**Pull request by deadwind4**
## What is the purpose of the change

* Support m1 chip.*

## Brief change log

  - *Update dependencies: numpy>= 1.21.4, pyarrow>=5.0.0, pandas>=1.3.0, apache-beam==2.38.0, cloudpickle==2.0.0, pemja== 0.1.5*

## Verifying this change

This change is already covered by existing tests, such as *(please describe tests)*.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (no)

### CODE REVIEW

File: flink-python/setup.py [hunk 1]
```diff
@@ -308,10 +308,10 @@ def extracted_output_files(base_dir, file_path, output_directory):
         author='Apache Software Foundation',
         author_email='dev@flink.apache.org',
         python_requires='>=3.6',
-        install_requires=['py4j==0.10.9.3', 'python-dateutil==2.8.0', 'apache-beam==2.27.0',
+        install_requires=['py4j==0.10.9.3', 'python-dateutil==2.8.0', 'apache-beam==2.36.0',
                           'cloudpickle==1.2.2', 'avro-python3>=1.8.1,!=1.9.2,<1.10.0',
-                          'pandas>=1.0,<1.2.0', 'pyarrow>=0.15.1,<3.0.0',
-                          'pytz>=2018.3', 'numpy>=1.14.3,<1.20', 'fastavro>=0.21.4,<0.24',
+                          'pandas>=1.3.0', 'pyarrow>=5.0.0',
+                          'pytz>=2018.3', 'numpy>=1.20.3', 'fastavro>=0.21.4,<0.24',
```
Comment (maver1ck) [inline flink-python/setup.py:314]: @deadwind4 
Can we upgrade numpy to >=1.21.0 ?
With 1.20.3 there are still problems with instalation on M1

Comment (deadwind4) [reply flink-python/setup.py:314]: OK, This Beam PR is also created by me :)

Comment (maver1ck) [reply flink-python/setup.py]: I know. Thanks.

**Comment by deadwind4** [comment]
@flinkbot run azure

**Comment by deadwind4** [comment]
@flinkbot run azure

**Comment by MartijnVisser** [comment]
@deadwind4 To fix the license build issues, you should edit `flink-python/src/main/resources/META-INF/NOTICE` and make the following changes:

- Replace `com.google.flatbuffers:flatbuffers-java:1.9.0` with `com.google.flatbuffers:flatbuffers-java:1.12.0`
- Remove `io.netty:netty-buffer:4.1.70.Final` and `io.netty:netty-common:4.1.70.Final` from line 13-14 since they also appear lower in the list with the other Netty dependencies
- Add `org.apache.arrow:arrow-memory-core:5.0.0` with the other Arrow dependencies
- Add `org.apache.beam:beam-vendor-bytebuddy-1_11_0:0.1` with the other Beam dependencies

**Comment by MartijnVisser** [comment]
@deadwind4 Will you have a look at the other CI reported failures?

**Comment by dianfu** [comment]
@deadwind4 What's the status of this PR?

**Comment by MartijnVisser** [comment]
@dianfu Based on the CI, it looks like there's a blocking situation on pemja since that requires numpy 1.19.5, while the other dependencies require higher versions:

```
Mar 14 09:47:15 The conflict is caused by:
Mar 14 09:47:15     apache-flink 1.15.dev0 depends on numpy>=1.21.4
Mar 14 09:47:15     apache-beam 2.36.0 depends on numpy<1.22.0 and >=1.14.3
Mar 14 09:47:15     pemja 0.1.2 depends on numpy==1.19.5
```

**Comment by deadwind4** [comment]
@dianfu I am waiting for pemja to release a new version, I will bump the pemja in this PR.
 https://github.com/alibaba/pemja/pull/1

**Comment by maver1ck** [comment]
@deadwind4 Pemja already upgraded Numpy.

**Comment by deadwind4** [comment]
> @deadwind4 Pemja already upgraded Numpy.

@maver1ck The newest Pemja release version has not upgraded the Numpy version.

**Comment by maver1ck** [comment]
> > @deadwind4 Pemja already upgraded Numpy.
> 
> @maver1ck The newest Pemja release version has not upgraded the Numpy version.

You're right. The numpy version bump wasn't merge to 0.1 branch :(

**Comment by MartijnVisser** [comment]
@deadwind4 Can I (try to) help with finalizing this PR? I can also test it on my M1 (and I think @maver1ck could also test it).

**Comment by deadwind4** [comment]
> @deadwind4 Can I (try to) help with finalizing this PR? I can also test it on my M1 (and I think @maver1ck could also test it).
@MartijnVisser Thank you a lot.
Yesterday, the pemja bumped to 0.1.5. I'm waiting for the mvn repo https://mvnrepository.com/artifact/com.alibaba/pemja . Now mvn repo has not 0.1.5, coming soon.

**Comment by MartijnVisser** [comment]
> Yesterday, the PemJa bumped to 0.1.5. I'm waiting for the mvn repo https://mvnrepository.com/artifact/com.alibaba/pemja . Now mvn repo has not 0.1.5. I talk with Xingbo Huang offline, it will be there in a few days.

Could it be that the Maven Repository website is outdated, because I do see it at https://repo1.maven.org/maven2/com/alibaba/pemja/ ?

**Comment by deadwind4** [comment]
> > Yesterday, the PemJa bumped to 0.1.5. I'm waiting for the mvn repo https://mvnrepository.com/artifact/com.alibaba/pemja . Now mvn repo has not 0.1.5. I talk with Xingbo Huang offline, it will be there in a few days.
> 
> Could it be that the Maven Repository website is outdated, because I do see it at https://repo1.maven.org/maven2/com/alibaba/pemja/ ?

@MartijnVisser you are right, I have rebased and pushed code, let's wait for CI.

**Comment by HuangXingBo** [comment]
I think we can bump apache-beam to the latest 2.38.0 (which provides the wheel package for python 3.9) and provide support for python 3.9. Support for python3.9 can also be split into another PR.

**Comment by HuangXingBo** [comment]
I have created the JIRA https://issues.apache.org/jira/browse/FLINK-27575 about support of Python 3.9.

**Comment by deadwind4** [comment]
We need to bump Beam to 2.39.0
https://github.com/PAIR-code/what-if-tool/issues/185#issuecomment-1013971242

File: NOTICE [hunk 2]
```diff
@@ -17,7 +17,7 @@ See bundled license files for details.
 This project bundles the following dependencies under the BSD license.
 See bundled license files for details.
 
-- cloudpickle:1.2.2
+- cloudpickle:2.0.0
```
Comment (HuangXingBo) [inline NOTICE:20]: We also need to replace the source package in the directory flink-python/lib

File: flink-python/setup.py [hunk 3]
```diff
@@ -310,12 +310,12 @@ def extracted_output_files(base_dir, file_path, output_directory):
...
-                          'cloudpickle==1.2.2', 'avro-python3>=1.8.1,!=1.9.2,<1.10.0',
-                          'pandas>=1.0,<1.2.0', 'pyarrow>=0.15.1,<3.0.0',
-                          'pytz>=2018.3', 'numpy>=1.14.3,<1.20', 'fastavro>=0.21.4,<0.24',
+        install_requires=['py4j==0.10.9.3', 'python-dateutil==2.8.0', 'apache-beam==2.38.0',
+                          'cloudpickle==2.0.0', 'avro-python3>=1.8.1,!=1.9.2,<1.10.0',
+                          'pandas>=1.3.0', 'pyarrow>=5.0.0',
+                          'pytz>=2018.3', 'numpy>=1.21.4', 'fastavro>=0.21.4,<0.24',
                           'requests>=2.26.0', 'protobuf<3.18',
-                          'pemja==0.1.4;python_full_version >= "3.7"',
+                          'pemja==0.1.5;python_full_version >= "3.7"', 'httplib2>=0.8,<0.19.0',
```
Comment (HuangXingBo) [inline flink-python/setup.py:318]: need to add `'httplib2>=0.8,<0.19.0'` to `dev-requirements.txt`?

**Review by HuangXingBo — COMMENTED** [review]
@deadwind4 Thanks a lot the PR. Good Job. I have only two minor comments.

**Review by HuangXingBo — APPROVED** [review]
LGTM

**Pull request closed without merging**